File: extract_pen_text.js
const fs = require('fs');

function extractText(node) {
    let texts = [];
    if (typeof node === 'object' && node !== null) {
        if (node.type === 'text' && node.content) {
            texts.push(node.content);
        }

        for (const key in node) {
            if (Object.prototype.hasOwnProperty.call(node, key)) {
                texts = texts.concat(extractText(node[key]));
            }
        }
    } else if (Array.isArray(node)) {
        for (const item of node) {
            texts = texts.concat(extractText(item));
        }
    }
    return texts;
}

try {
    const data = fs.readFileSync('pencil-welcome.pen', 'utf8');
    const json = JSON.parse(data);
    const texts = extractText(json);
    console.log(texts.join('\n'));
} catch (e) {
    console.error(e);
}

File: idea-dump/scripts/extract_pen_text.py
import json
import sys

def extract_text(node):
    texts = []
    if isinstance(node, dict):
        if node.get('type') == 'text' and 'content' in node:
            texts.append(node['content'])
        
        for key, value in node.items():
            texts.extend(extract_text(value))
    elif isinstance(node, list):
        for item in node:
            texts.extend(extract_text(item))
    return texts

try:
    with open('pencil-welcome.pen', 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    texts = extract_text(data)
    for t in texts:
        print(t)
except Exception as e:
    print(f"Error: {e}")
